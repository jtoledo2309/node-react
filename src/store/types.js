export const AUTH_LOGIN_REQUEST = "AUTH_LOGIN_REQUEST";
export const AUTH_LOGIN_SUCESS = "AUTH_LOGIN_SUCESS";
export const AUTH_LOGIN_FAILURE = "AUTH_LOGIN_FAILURE";

export const AUTH_LOGOUT = "AUTH_LOGOUT";

export const ADVERTS_LOADED = "ADVERTS_LOADED";
export const ADVERT_CREATED = "ADVERT_CREATED";

export const UI_RESET_ERROR = "UI_RESET_ERROR";
