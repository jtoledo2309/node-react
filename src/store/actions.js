import {
  AUTH_LOGOUT,
  ADVERTS_LOADED,
  AUTH_LOGIN_SUCESS,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_FAILURE,
  UI_RESET_ERROR,
} from "./types.js";

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSucess = () => ({
  type: AUTH_LOGIN_SUCESS,
});

export const authLoginFailure = (error) => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const advertsLoaded = (adverts) => ({
  type: ADVERTS_LOADED,
  payload: adverts,
});

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
