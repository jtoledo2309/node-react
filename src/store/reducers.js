// auth: true/false
// adverts: []
// ui: {
//   isLoading: true/false
//   error: error/null
// }

import {
  AUTH_LOGIN_SUCESS,
  AUTH_LOGOUT,
  ADVERTS_LOADED,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_FAILURE,
  UI_RESET_ERROR,
} from "./types.js";

const defaultState = {
  auth: false,
  adverts: [],
  ui: {
    isLoading: false,
    error: null,
  },
};

// export default function reducer(state = defaultState, action) {
//   switch (action.type) {
//     case AUTH_LOGIN:
//       return { ...state, auth: true };

//     case AUTH_LOGOUT:
//       return { ...state, auth: false };

//     case ADVERTS_LOADED:
//       return { ...state, adverts: action.payload };

//     default:
//       return state;
//   }
// }

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_SUCESS:
      return true;

    case AUTH_LOGOUT:
      return false;

    default:
      return state;
  }
}

export function adverts(state = defaultState.adverts, action) {
  if (action.type === ADVERTS_LOADED) {
    return action.payload;
  }
  return state;
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return {
      isLoading: false,
      error: action.payload,
    };
  }
  if (/_REQUEST$/.test(action.type)) {
    return {
      error: null,
      isLoading: true,
    };
  }
  if (/SUCESS$/.test(action.type)) {
    return {
      error: null,
      isLoading: false,
    };
  }
  if (action.type === UI_RESET_ERROR) {
    return {
      ...state,
      error: null,
    };
  }
  return state;
}
// switch (action.type) {
//   case AUTH_LOGIN_REQUEST:
//     return {
//       error: null,
//       isLoading: true,
//     };
//   case AUTH_LOGIN_SUCESS:
//     return {
//       error: null,
//       isLoading: false,
//     };
//   case AUTH_LOGIN_FAILURE:
//     return {
//       error: action.payload,
//       isLoading: false,
//     };
//   case UI_RESET_ERROR:
//     return {
//       ...state,
//       error: null,
//     };
//   default:
//     return state;
// }
// export default function reducer(state = defaultState, action) {
//   return {
//     auth: auth(state.auth, action),
//     adverts: adverts(state.adverts, action),
//   };
// }
