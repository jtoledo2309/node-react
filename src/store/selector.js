export const getIsLogged = (state) => state.auth;

export const getAdverts = (state) =>
  state.adverts.areLoaded ? state.adverts.data : [];

export const areAdvertsLoaded = (state) => state.adverts.areLoaded;

export const getUniqueAdvert = (advertId) => (state) =>
  state.adverts.data.find((advert) => advert.id.toString() === advertId);

export const getUi = (state) => state.ui;
