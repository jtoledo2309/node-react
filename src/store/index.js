import { createStore, combineReducers } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";
import * as reducers from "./reducers";

const reducer = combineReducers(reducers);
export default function configureStore(preloadedState) {
  const store = createStore(reducer, preloadedState, composeWithDevTools());

  return store;
}
