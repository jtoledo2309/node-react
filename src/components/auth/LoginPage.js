import { useState } from "react";
import Button from "../common/Button";
import FormField from "../common/FormField";
import { useLocation, useNavigate } from "react-router-dom";

import "./LoginPage.css";
import CheckBox from "../common/Checkbox";
import { useDispatch, useSelector } from "react-redux";
import {
  authLoginSet,
  uiResetError,
  authLoginNotSet,
} from "../../store/actions";
import { getUi } from "../../store/selector";

const LoginPage = ({ onLogin }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const location = useLocation();
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);
  const [rememberLogin, setRememberLogin] = useState(false);
  const handleResetError = () => dispatch(uiResetError());

  const handleChangeUsername = (event) => setEmail(event.target.value);
  const handleChangePassword = (event) => setPassword(event.target.value);

  const handleRememberLogin = (event) => {
    event.target.checked ? setRememberLogin(true) : setRememberLogin(false);
  };

  const handleSubmit = async (event) => {
    event.preventDefault();

    //dispatch(authLoginRequest());
    if (rememberLogin) {
      await dispatch(authLoginSet({ email, password }));
      //await login({ email, password });
      //dispatch(authLoginSucess());
      const previousroute = location.state?.from?.pathname || "/";
      navigate(previousroute, { replace: true });
    } else {
      await dispatch(authLoginNotSet({ email, password }));
      //await loginNotSet({ email, password });
      //dispatch(authLoginSucess());
      const previousroute = location.state?.from?.pathname || "/";
      navigate(previousroute, { replace: true });
    }
  };

  const isDisabled = () => {
    return !(email && password && !isLoading);
  };
  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in to Node-React</h1>
      <form onSubmit={handleSubmit}>
        <FormField
          type="text"
          name="email"
          label="Enter your email"
          className="loginForm-field"
          onChange={handleChangeUsername}
          value={email}
        />
        <FormField
          type="password"
          name="password"
          label="Enter your password"
          className="loginForm-field"
          onChange={handleChangePassword}
          value={password}
        />
        <CheckBox
          type="checkbox"
          name="remember-password"
          label="Quiero recordar tu usuario para proximas sesiones"
          onChange={handleRememberLogin}
          value={rememberLogin}
        />
        <Button
          type="submit"
          variant="primary"
          className="loginForm-submit"
          disabled={isDisabled()}
        >
          LogIn
        </Button>
      </form>

      {error && (
        <div className="loginPage-error" onClick={handleResetError}>
          {error.message}
        </div>
      )}
    </div>
  );
};

export default LoginPage;
