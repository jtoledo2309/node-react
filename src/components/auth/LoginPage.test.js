import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { defaultState } from "../../store/reducers";
import LoginPage from "./LoginPage";
import { BrowserRouter as Router } from "react-router-dom";
import { authLoginNotSet } from "../../store/actions";

jest.mock("../../store/actions");

describe("loginPage", () => {
  const store = {
    getState: () => defaultState,
    dispatch: () => {},
    subscribe: () => {},
    location: () => {},
    navigate: () => {},
  };

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <Router>
          <LoginPage />
        </Router>
      </Provider>
    );

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("deberia lanzar el authloginNotSet", () => {
    const email = "prueba";
    const password = "1234";

    renderComponent();

    const usernameInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole("button");

    expect(submitButton).toBeDisabled();
    fireEvent.change(usernameInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });
    expect(submitButton).toBeEnabled();

    fireEvent.click(submitButton);
    expect(authLoginNotSet).toHaveBeenCalledWith({ email, password });
  });
});
