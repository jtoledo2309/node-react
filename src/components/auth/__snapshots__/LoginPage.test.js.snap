// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to Node-React
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            Enter your email
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            Enter your password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <div
        class="checkbox-general"
      >
        <input
          name="remember-password"
          type="checkbox"
          value="false"
        />
        <label>
           
          Quiero recordar tu usuario para proximas sesiones
        </label>
      </div>
      <button
        class="sc-bcXHqe diyiTl loginForm-submit"
        disabled=""
        type="submit"
      >
        LogIn
      </button>
    </form>
  </div>
</div>
`;
